## Fråga 10

### Arbeta med `salary_dataset.csv` och jämför två regressionsmodeller med korsvalidering

**Svar:**

Datasetet innehåller `YearsExperience` och `Salary`. Eftersom målet är att förutsäga lön använder jag `Salary` som den beroende variabeln `y`, medan `YearsExperience` används som `X`. Alltså är Salary våran target och YearsExperience våran feature.


In [1]:
# Imports

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, cross_validate
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import root_mean_squared_error

In [2]:
# Läser in vår dataframe
df = pd.read_csv("chapter_02_ml_project\data\salary_dataset (1).csv")

<>:2: SyntaxWarning: invalid escape sequence '\d'
<>:2: SyntaxWarning: invalid escape sequence '\d'
C:\Users\chris\AppData\Local\Temp\ipykernel_23036\2127734866.py:2: SyntaxWarning: invalid escape sequence '\d'
  df = pd.read_csv("chapter_02_ml_project\data\salary_dataset (1).csv")


In [3]:
# Definerar våran feature och target
X = df[["YearsExperience"]]
y = df["Salary"]

In [4]:
# Delar upp vår data i tränings och testdata

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

In [5]:
# Kollar 5 första kolumnerna 
df.head()

,YearsExperience,Salary
0,1.2,39344.0
1,1.4,46206.0
2,1.6,37732.0
3,2.1,43526.0
4,2.3,39892.0


In [6]:
# Träna två regressionsmodeller på träningsdatan med 5-delad korsvalidering. Använder neg RMSE som scoring

linear_model = LinearRegression()
tree_model = DecisionTreeRegressor(random_state=42)

linear_cv = cross_validate(
    linear_model,
    X_train,
    y_train,
    cv=5,
    scoring="neg_root_mean_squared_error"
)

tree_cv = cross_validate(
    tree_model,
    X_train,
    y_train,
    cv=5,
    scoring="neg_root_mean_squared_error"
)

In [7]:
# Här jämför vi modellernas resultat

linear_rmse = -linear_cv["test_score"].mean()
tree_rmse = -tree_cv["test_score"].mean()

print("Linear Regression CV RMSE:", linear_rmse)
print("Decision Tree CV RMSE:", tree_rmse)

Linear Regression CV RMSE: 5293.203196998775
Decision Tree CV RMSE: 5611.57858424216


Linear Regression fick lägst RMSE i korsvalideringen och väljs därför för den slutliga utvärderingen på testdatan.

In [8]:
from sklearn.metrics import root_mean_squared_error

linear_model.fit(X_train, y_train)

y_pred = linear_model.predict(X_test)

test_rmse = root_mean_squared_error(y_test, y_pred)

print("Test RMSE:", test_rmse)

Test RMSE: 7059.043621901506


Linear Regression presterade bäst i korsvalideringen med ett RMSE på cirka `5293`, jämfört med Decision Tree som fick cirka `5612`.

På testdatan fick Linear Regression ett RMSE på cirka `7059`.

Det innebär att modellens prediktioner i genomsnitt ligger ungefär några tusen enheter från de faktiska lönerna. Inte så bra men inte så tokigt heller!